src/transformers/models/big_bird/modeling_big_bird.py

```diff
@@ -0,0 +1,2695 @@
+            dim=2,
+        )
+        context_layer = context_layer.view((b, h, m, -1)) * from_mask
+        context_layer = torch.transpose(context_layer, 1, 2)
+
+        if output_attentions:
+            # TODO(PVP): need to verify if below code is correct
+            attention_probs = torch.zeros(b, h, m, n, dtype=torch.float, device=context_layer.device)
+
+            # corresponding to `first_context_layer`
+            attention_probs[:, :, :wm, :] = first_attn_weights
+
+            # corresponding to `second_context_layer`
+            attention_probs[:, :, wm : 2 * wm, : 3 * wn] = second_attn_weights[:, :, :, : 3 * wn]
+            attention_probs[:, :, wm : 2 * wm, -wn:] = second_attn_weights[:, :, :, 3 * wn : 4 * wn]
+            for p1, i1, w1 in zip(range(b), rand_attn, second_attn_weights):
+                for p2, i2, w2 in zip(range(h), i1, w1):
+                    attention_probs.view(b, h, m // wm, wm, n // wn, wn)[p1, p2, 1, :, i2[0]] = w2[:, 4 * wn :].view(
+                        wm, r, wn
+                    )
+
+            # corresponding to `context_layer`
+            for q_idx in range(m // wm - 4):
+                slice = attention_probs.view(b, h, m // wm, wm, n // wn, wn)[:, :, 2:-2, :, 1:-1, :]
+                slice[:, :, q_idx, :, q_idx : q_idx + 3, :] = attn_weights[:, :, q_idx, :, wn : 4 * wn].view(
+                    b, h, wm, 3, wn
+                )  # inner_band_product
+            attention_probs[:, :, 2 * wm : -2 * wm, :wn] = attn_weights[:, :, :, :, :wn].view(
+                b, h, -1, wn
+            )  # first_band_product
+            attention_probs[:, :, 2 * wm : -2 * wm, -wn:] = attn_weights[:, :, :, :, -wn:].view(
+                b, h, -1, wn
+            )  # last_band_product
+            for p1, i1, w1 in zip(range(b), rand_attn, attn_weights):
+                for p2, i2, w2 in zip(range(h), i1, w1):
+                    for q_idx in range(1, len(i2) - 1):
+                        attention_probs.view(b, h, m // wm, wm, n // wn, wn)[p1, p2, q_idx + 1, :, i2[q_idx]] = w2[
+                            q_idx - 1, :, 4 * wn : -wn
+                        ].view(wm, r, wn)
+
+            # corresponding to `second_last_context_layer`
+            attention_probs[:, :, -2 * wm : -wm, :wn] = second_last_attn_weights[:, :, :, :wn]
+            attention_probs[:, :, -2 * wm : -wm, -3 * wn :] = second_last_attn_weights[:, :, :, wn : 4 * wn]
+            for p1, i1, w1 in zip(range(b), rand_attn, second_last_attn_weights):
+                for p2, i2, w2 in zip(range(h), i1, w1):
+                    attention_probs.view(b, h, m // wm, wm, n // wn, wn)[p1, p2, -2, :, i2[-1]] = w2[:, 4 * wn :].view(
+                        wm, r, wn
+                    )
+
+            # corresponding to `last_context_layer`
+            attention_probs[:, :, -wm:, :] = last_attn_weights
+
+        else:
+            attention_probs = None
+
+        return context_layer, attention_probs
```

Comment:
This is really nice already! It's probably the most complicated attention layer we have in Transformers. 
If possible could you maybe add some more comments that somewhat explain what is happening, like "The following code applies random attention", the following block applies global attention, ... (I know it's a lot to ask for and you've already added a bunch of very nice comments, but given the complexity of this function, it would be awesome we could add a couple more comments)